### PR TITLE
feat: expose wrapped (Fast) React components

### PR DESCRIPTION
**Pull request by mattrothenberg**
### Link to relevant issue
This pull request resolves https://github.com/microsoft/vscode-webview-ui-toolkit/issues/299

### Description of changes

This PR adds the ability for users to import wrapped React components directly from a subdirectory of the `@vscode/webview-ui-toolkit` package thusly:

```tsx
import { VSCodeButton, VSCodeCheckbox } from '@vscode/webview-ui-toolkit/react'
```
Specifically, this PR:

**Adds a `src/react/index.ts` file which exports wrapped versions of all of the components in this library**

```tsx
import React from 'react';
import {provideReactWrapper} from '@microsoft/fast-react-wrapper';
import {provideFASTDesignSystem} from '@microsoft/fast-components';

import { vsCodeBadge } from '..'

const {wrap} = provideReactWrapper(React, provideFASTDesignSystem());

export const VSCodeBadge = wrap(vsCodeBadge());
// ETCETERA
```

**Adds a `tsconfig-react.json` and (supporting NPM script) to package up the aforementioned file.**

```
// package.json
"build:react": "tsc -p ./tsconfig-react.json",
```

**Proposes a modification to our build process, such that users can import directly from the `react` subdirectory. 🚨 Please note that this requires a significant change to the build process. 🚨**

Specifically, the steps are:
1. After building both the regular components _and_ the React wrapped components, copying a stub version of the `package.json` file to the `/dist` directory. This stub file is a mirror image of the original `package.json` with the exception of the values in the `main` and `types` fields losing their `/dist` prefix.
2.`cd`ing into the `/dist` directory and running publish from within there. This ensures that the built files that end up being distributed on NPM have a flat structure (rather than a nested dist directory) 👇 

**Old, Nested Directory structure**
```
@vscode/webview-ui-toolkit/
├─ dist/
│  ├─ esm/
│  ├─ dst/
├─ package.json
```
In this old world, the `main` file was `/dist/esm/index.js`.

**New, Flattened Directory structure**
```
@vscode/webview-ui-toolkit/
├─ package.json
├─ react/
├─ dst/
├─ esm/
```
In this new world, the `main` file is `/esm/index.js`, but users can also import directly from `react/index.js`.

There are likely less kludgy ways of doing this, but this seems to work 🤷 .

Finally, I'd like to follow-up this PR by building a Storybook dedicated solely to the wrapped React components. As far as I an tell, it's extremely difficult (if not impossible) to use multiple frameworks within a single Storybook instance. It is possible, however, to "compose" storybooks via the following API https://storybook.js.org/docs/react/workflows/storybook-composition. To that end, once this PR is merged and once I have the other Storybook set up, I will update the Storybook here to pull in the published React SB instance.

### CODE REVIEW

**Comment by mattrothenberg** [comment]
cc @hawkticehurst ! I wasn't able to add you as a reviewer

File: package.json
Comment (hawkticehurst) [inline package.json:40]: Realized this over the holiday weekend, but could we do some testing to see if it's possible to declare React as a peer dependency (or something along those lines) so the version of React that is being used by an extension author is automatically picked up/used by these components?

Basically, I would hope to avoid the scenario where someone who is using React `v17+` can't use these components because we've shipped the components using `v16`.

Comment (hawkticehurst) [reply package.json:40]: I'm not entirely sure that declaring react as a peer dependency is the solution either, but seems like it might be a good starting point to explore.

Comment (hawkticehurst) [reply package.json:40]: This is what `fast-react-wrapper` does as a reference point:

https://github.com/microsoft/fast/blob/master/packages/utilities/fast-react-wrapper/package.json#L91-L93

File: src/react/index.ts
Comment (hawkticehurst) [inline src/react/index.ts:3]: The toolkit actually exports its own `provideVSCodeDesignsystem` from `src/index.ts` that should be used here instead of the FAST design system.

```typescript
import {provideVSCodeDesignSystem} from '../index';
```

File: src/react/index.ts
Comment (hawkticehurst) [inline src/react/index.ts:27]: Update to use `provideVSCodeDesignsystem` instead of `provideFASTDesignSystem`.

File: src/react/index.ts
Comment (hawkticehurst) [inline src/react/index.ts:25]: We have an eslint config that enforces alphabetical imports which is why you're getting a build error here.

You can use `npm run lint:fix` to automatically fix this error locally then push up the changes.

Comment (hawkticehurst) [reply src/react/index.ts:25]: Also have `npm run fmt:fix` for any formatting errors you may run into as an FYI.

File: src/react/index.ts
Comment (hawkticehurst) [inline src/react/index.ts:47]: If you could alphabetically order all of these exports that would be appreciated too just to match the conventions of the rest of the toolkit.

File: tsconfig-react.json
Comment (hawkticehurst) [inline tsconfig-react.json:14]: Super minor nit, but could you rename this file as `tsconfig.react.json` to match the convention already established by `tsconfig.eslint.json`?

File: package.json
Comment (hawkticehurst) [inline package.json:36]: You can uninstall the `@microsoft/fast-components` package.

It's only used to get `provideFASTDesignsystem` right now (which I've mentioned in other comments can be replaced with `provideVSCodeDesignsystem`).

File: package.json
Comment (hawkticehurst) [inline package.json:22]: It seems like `copy-manifest` is defined here but never used? If that is the case could we make sure to remove it before merging this PR?

File: copy-pkg-json.js
Comment (hawkticehurst) [inline copy-pkg-json.js:12]: This is just a personal preference of mine, but could `copy-pkg-json.js` be put inside a new `scripts` directory that can be created/placed in the root directory, just to keep things a bit more tidy/separated from the toolkit config?

**Review by hawkticehurst — CHANGES_REQUESTED** [review]
Finally had a free moment to take a closer pass on this PR and it's looking great! 

Left a couple of comments asking for changes and some minor nits when you're back in the office, but super excited with the progress! Thank you again for the help.

File: copy-pkg-json.js
Comment (hawkticehurst) [reply copy-pkg-json.js:12]: Oh and then you'll also want to update `package.json` build script to reflect the new path to `copy-pkg-json.js`:

```json
"build": "rollup -c && tsc -p ./tsconfig.json && npm run doc && npm run build:react && ./scripts/copy-pkg-json.js",
```

File: package.json
Comment (mattrothenberg) [reply package.json]: resolved by f8be1f489161e1a9bd15a0acbed9ec09628f4ece

File: src/react/index.ts
Comment (mattrothenberg) [reply src/react/index.ts]: Resolved by 82e710a8991eebce74fdb702c03f81171c6b684d

File: src/react/index.ts
Comment (mattrothenberg) [reply src/react/index.ts]: Resolved by 82e710a8991eebce74fdb702c03f81171c6b684d

File: src/react/index.ts
Comment (mattrothenberg) [reply src/react/index.ts:47]: resolved by 15c25102230abfd0e64c48c89c8d5d98b070c6c1

File: src/react/index.ts
Comment (mattrothenberg) [reply src/react/index.ts]: resolved by 15c25102230abfd0e64c48c89c8d5d98b070c6c1

File: package.json
Comment (mattrothenberg) [reply package.json]: Resolved by 82e710a8991eebce74fdb702c03f81171c6b684d

File: tsconfig-react.json
Comment (mattrothenberg) [reply tsconfig-react.json]: resolved by 0929fbbc93f8732e020c755d04d07a132ee493ac

File: package.json
Comment (mattrothenberg) [reply package.json:22]: This is what I wanted to discuss with you, as this script has to be run at publish time in order to create the subdirectories for the `/react` imports

File: copy-pkg-json.js
Comment (mattrothenberg) [reply copy-pkg-json.js]: Resolved by a25ce7b

File: package.json
Comment (mattrothenberg) [reply package.json]: Resolved by 2b0eb922b5eff7d077c023265d52308af1554454

**Comment by mattrothenberg** [comment]
@hawkticehurst All feedback addressed! I've tested that the build works locally, though I'm confused about one thing.

I've been testing locally with a tool called [`yalc`](https://github.com/wclr/yalc) by doing the following:
- Running `npm run build`
- `cd`ing into the `dist` directory
- Running `yalc publish`

(You'll recall that in this new world, we have to publish **from within** the dist directory otherwise we won't get the subdirectory imports. I wish there was a better way of doing this 🤔 )

Unfortunately the latter `yalc publish` command consistently fails when the `prepublishOnly` script is present in the project's `package.json`, since that script triggers the build command again and fails to find the `rollup.config.js` file (since we're in `dist` at this point, not root).

Any thoughts here?

**Comment by hawkticehurst** [comment]
> Any thoughts here?

Hmm, it's probably just a case of removing the `prepublishOnly` script for the time being (with the assumption that at some point TypeScript will eventually ship the feature that would make exporting subdirectory type info way easier) and manually run `npm run build` before I publish. I only had the pre-publish script as a contingency in case I ever forgot to build before publishing but I've done pretty well so far.

With that said, I could envision potentially doing something fancy with the npm `--prefix` flag, so I can do some experiments today. And in the meantime, I think removing the `prepublishOnly` script is fine (assuming that fixes the problem).

**Comment by hawkticehurst** [comment]
> (You'll recall that in this new world, we have to publish **from within** the dist directory otherwise we won't get the subdirectory imports. I wish there was a better way of doing this 🤔 )

Lol thanks for the reminder, I had forgotten about that. But yeah we can manage for now and hopefully the ability to export subdirectory type info will land sooner versus later 😪.

File: src/react/index.ts
Comment (hawkticehurst) [inline src/react/index.ts:47]: As I was checking out this code locally I just realized: Can we add TSDoc comments on all of these exports so that there's a little more intellisense info for those importing the components?

Adding comments for each component with the below format would be perfect for now. Thanks!

```typescript
/**
 * VS Code Badge React component.
 *
 * @public
 */
export const VSCodeBadge = wrap(vsCodeBadge());
```

Comment (hawkticehurst) [inline src/react/index.ts:47]: I also just remembered that we still need to configure component events for most of these. Have you done any work on that yet?

https://www.fast.design/docs/integrations/react/#configuring-custom-events

**Review by hawkticehurst — COMMENTED** [review]
Changes look and function great! Thank you!

While reviewing I did catch two more things that I left comments on.

File: src/react/index.ts
Comment (mattrothenberg) [reply src/react/index.ts:47]: Sure, I'll tackle this tomorrow.

Comment (mattrothenberg) [reply src/react/index.ts:47]: I have not. I assumed we were still blocked on this, considering native DOM events don't even seem to be working. For example 👇 

```ts
<VSCodeCheckbox
  checked // TypeError, also doesn't affect the state of the checkbox 
  onChange={(e) => {
    console.log(e); // never fires
  }}
>
  Subscribe
</VSCodeCheckbox>
```

Comment (hawkticehurst) [reply src/react/index.ts:47]: Wonderful thank you!

Comment (hawkticehurst) [reply src/react/index.ts:47]: Yep sounds great, we can hold off on this until a resolution to [this issue](https://github.com/microsoft/fast/issues/5387) comes down the line.

Comment (mattrothenberg) [reply src/react/index.ts]: Resolved by 0b68e85

**Comment by pwang347** [comment]
@hawkticehurst Hi Hawk! I read through the discussions in the fast repo, and was just wondering if there's an ETA on getting the stop-gap solution merged. We're currently hoping to adopt the toolkit (in React) to try it out as early as possible.

**Comment by hawkticehurst** [comment]
@pwang347 Hi Paul! I'd wager publishing a set of React components (i.e. this PR) is probably still two or so weeks out. 

However, if you're **_really eager and can't wait_** to get up and running I could decidedly have the stop-gap solution mentioned in those issue threads published first thing next week, which would unblock the ability for toolkit users to wrap the web components themselves (I would, of course, give some guidance on how to do that properly).

Then once this PR lands (which will do the job of wrapping components for toolkit users) you could simply remove your wrapping code and switch to importing React components directly.

Lemme know if that's something you would be interested in or if you'd rather wait for this PR. :)

**Comment by pwang347** [comment]
@hawkticehurst Sounds good! Yes, I think even with the added manual wrapping, having the stopgap solution in the meantime to unblock would be great (if it's not too much trouble for you). Appreciate the fast response, thanks!

**Comment by hawkticehurst** [comment]
> @hawkticehurst Sounds good! Yes, I think even with the added manual wrapping, having the stopgap solution in the meantime to unblock would be great (if it's not too much trouble for you). Appreciate the fast response, thanks!

The temporary solution has been published and a quick summary/guide on the update has been posted [here](https://github.com/microsoft/vscode-webview-ui-toolkit/issues/283#issuecomment-1005310954) (decided to put it in a thread with a higher visibility conversation about the ongoing React work).

Hope that helps unblock you and feel free to open GH issues with any questions or problems you run into!

**Comment by pwang347** [comment]
@hawkticehurst this is great, will give it a try today or tomorrow! thanks again!

**Comment by mattrothenberg** [comment]
Latest changes from `main` have been merged in 👍

**Comment by hawkticehurst** [comment]
> @hawkticehurst this is great, will give it a try today or tomorrow! thanks again!

Great, let us know how it goes!

**Comment by pwang347** [comment]
> > @hawkticehurst this is great, will give it a try today or tomorrow! thanks again!
> 
> Great, let us know how it goes!

Gave it a try, and so far everything seems to be working great (plugged it into the app, events were working and things felt good)! Only issue I ran into was https://github.com/microsoft/fast/issues/5198, but was able to resolve it with the `skipLibCheck` workaround.

**Comment by hawkticehurst** [comment]
> Gave it a try, and so far everything seems to be working great (plugged it into the app, events were working and things felt good)! Only issue I ran into was https://github.com/microsoft/fast/issues/5198, but was able to resolve it with the `skipLibCheck` workaround.

That's great to hear! Also wasn't aware of this issue so thanks for mentioning it/giving a heads up!

**Comment by hawkticehurst** [comment]
> Also wasn't aware of this issue so thanks for mentioning it/giving a heads up!

After reading through, I realize I _was_ actually aware of this so thank you for the reminder!

**Comment by hawkticehurst** [comment]
And just to be super clear in case others run into this problem--you got this issue because you were using a TS version >4.3.5?

**Comment by pwang347** [comment]
> And just to be super clear in case others run into this problem--you got this issue because you are using a TS version >4.3.5?

Apologies for the late reply, yes I am using 4.4.4!

**Comment by hawkticehurst** [comment]
> Apologies for the late reply, yes I am using 4.4.4!

No worries at all! Thanks for clarifying!

File: src/react/index.ts
Comment (hawkticehurst) [inline src/react/index.ts:91]: Nice, thank you!

**Comment by mattrothenberg** [comment]
Names added via 3763b5e

**Review by hawkticehurst — APPROVED** [review]
Looks fantastic, left some notes/requests for final polish but I think we're basically there!

File: scripts/copy-pkg-json.js
Comment (mattrothenberg) [reply scripts/copy-pkg-json.js]: Resolved by c85751d

File: scripts/copy-pkg-json.js
Comment (mattrothenberg) [reply scripts/copy-pkg-json.js]: Resolved by 5e284e3

File: src/react/index.ts
Comment (mattrothenberg) [reply src/react/index.ts]: Resolved by 7ce6c9f

File: src/react/index.ts
Comment (mattrothenberg) [reply src/react/index.ts]: Resolved by 320beca

**Comment by hawkticehurst** [comment]
Incredible work @mattrothenberg! Thank you again for helping make this a reality!

**Comment by hawkticehurst** [comment]
@pwang347 I'm sure you probably saw already, but in case you didn't as of `v0.9.0` we finally shipped the React components if you would like to use them.

You should be able to remove the file that wraps components (and uninstall the `fast-react-wrapper` package) and instead replace it with import statements along these lines and it should just work:

```ts
import { VSCodeButton, VSCodeDropdown } from '@vscode/webview-ui-toolkit/react';
```

Let me know if you have any questions!

**Comment by pwang347** [comment]
@hawkticehurst This is great, thank you!

**Pull request merged**
